### DISCUSSION TITLE
Training on Intel Arc, GradScaler missing...

### DISCUSSION THREAD

**Discussion by z0mbi32** [vladmandic/sdnext#1334] [General]
Trying to train a model, but this is the last thing I'm seeing in the console:

```
╯
AttributeError: module 'intel_extension_for_pytorch.xpu.amp' has no attribute
'GradScaler'

``` 

Is this just that GradScaler isn't included in that module, so I'm SOL for training with OneAPI at the moment?

**Comment by Disty0** [comment]
GradScaler for GPU doesn't exist with IPEX.
Tried to replace it with the CPU one but it's still broken for now.

**Comment by Disty0** [comment]
Removed GradScaler for ipex altogether.
Training should work with BF16 now.

Also don't go over Batch Size 4 unless you want to train a noise generator.

**Comment by Disty0** [comment]
It wasn't really working without GradScaler so apply this patch to use it:
https://github.com/Disty0/ipex_grad_scaler

Go to the Compute Settings and change "Device precision type" to BF16.
Then patch the XPU AMP with this command and restart the WebUI:
Note: Run this command in the automatic folder.
``get https://raw.githubusercontent.com/Disty0/ipex_grad_scaler/main/patch.sh && chmod +x patch.sh && ./patch.sh``

**Reply by z0mbi32** [reply]
I'm trying the patch but getting this at the end..

```
mv: cannot move '__init__.py' to '../venv/lib/python3.10/site-packages/intel_extension_for_pytorch/xpu/amp/__init__.py': No such file or directory
mv: cannot move 'grad_scaler.py' to '../venv/lib/python3.10/site-packages/intel_extension_for_pytorch/xpu/amp/grad_scaler.py': No such file or directory
rmdir: failed to remove 'patch': Directory not empty
```

Then still getting the gradscaler in the console when trying to train.

**Reply by z0mbi32** [reply]
To follow on this, I beleive my equivalent folders look like:

`/usr/local/lib/python3.10/dist-packages/intel_extension_for_pytorch/`

I've tried manually moving the patched files to no avail

**Reply by Disty0** [reply]
Automatic uses it's own virtual environment. Run this patch in the automatic folder.
If you want to use this patch for different applications, manually replace the files inside of the xpu/amp folder in the intel_extension_for_pytorch library.
